src/api/app/models/bs_request.rb

```diff
@@ -1033,4 +1052,180 @@ def expand_targets
+    end
+    roles = opts[:roles] || []
+    states = opts[:states] || []
+
+    # it's wiser to split the queries
+    if opts[:project] && roles.empty? && (states.empty? || states.include?('review'))
+      rel = collection(opts.merge(roles: %w(reviewer)))
+      ids = rel.ids
+      rel = collection(opts.merge(roles: %w(target source)))
+    else
+      rel = collection(opts)
+      ids = []
+    end
+    ids.concat(rel.ids)
+  end
+
+  private
+
+  def self.extend_query_for_group(group, requests, roles, review_states)
+    inner_or = []
+    group = Group.find_by_title!(group)
+
+    # find requests where group is maintainer in target project
+    requests, inner_or = extend_query_for_maintainer(group, requests, roles, inner_or)
+
+    if roles.empty? || roles.include?('reviewer')
+      requests = requests.includes(:reviews).references(:reviews)
+      # requests where the user is reviewer or own requests that are in review by someone else
+      or_in_and = %W(reviews.by_group=#{quote(group.title)})
+
+      requests, inner_or = extend_query_for_involved_reviews(group, or_in_and, requests, review_states, inner_or)
+    end
+    if inner_or.empty?
+      requests.none
+    else
+      requests.where(inner_or.join(' or '))
+    end
+  end
+
+  def self.extend_query_for_user(user, requests, roles, review_states, package, subprojects, project)
+    inner_or = []
+    user = User.find_by_login!(user)
+
+    # user's own submitted requests
+    if roles.empty? || roles.include?('creator')
+      inner_or << "bs_requests.creator = #{quote(user.login)}"
+    end
+    # find requests where user is maintainer in target project
+    requests, inner_or = extend_query_for_maintainer(user, requests, roles, inner_or)
+    if roles.empty? || roles.include?('reviewer')
+      requests = requests.includes(:reviews).references(:reviews)
+
+      # requests where the user is reviewer or own requests that are in review by someone else
+      or_in_and = %W(reviews.by_user=#{quote(user.login)})
+
+      # include all groups of user
+      usergroups = user.groups.map { |group| "'#{group.title}'" }
+      or_in_and << "reviews.by_group in (#{usergroups.join(',')})" unless usergroups.blank?
+
+      requests, inner_or = extend_query_for_involved_reviews(user, or_in_and, requests, review_states, inner_or)
+    end
+    if inner_or.empty?
+      requests.none
+    else
+      requests.where(inner_or.join(' or '))
+    end
+  end
+
+  def self.extend_query_for_project(requests, roles, states, review_states, package, subprojects, project)
+    inner_or = []
+    requests, inner_or = extend_relation('source', requests, roles, package, subprojects, project, inner_or)
+    requests, inner_or = extend_relation('target', requests, roles, package, subprojects, project, inner_or)
+
+    if (roles.empty? || roles.include?('reviewer')) &&
+       (states.empty? || states.include?('review'))
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.